### PR TITLE
Miscellaneous python typing fixes

### PR DESCRIPTION
**Pull request by marcospri**
The python typecheker (mypy) can be run with `make typecheck`. It's not enabled in either `make sure` or in CI.

Fix some low-hanging-fruit issues and add a few `ignores` in more complicated ones.

### CODE REVIEW

File: lms/models/application_instance.py [hunk 1]
```diff
@@ -80,7 +80,7 @@ class ApplicationInstance(CreatedUpdatedMixin, Base):
     organization_id = sa.Column(
         sa.Integer(), sa.ForeignKey("organization.id"), nullable=True
     )
-    organization = sa.orm.relationship("Organization")
+    organization: sa.Mapped["Organization"] = sa.orm.relationship("Organization")
```
Comment (marcospri) [inline lms/models/application_instance.py:83]: We'll eventually should type all the models as they are at the core of the codebase.

File: lms/product/plugin/plugin.py [hunk 2]
```diff
@@ -31,9 +31,9 @@ def __get__(self, instance, owner):
             setattr(instance, self.plugin_name, plugin)  # Overwrite the attr
             return plugin
 
-    grouping: GroupingPlugin = _LazyPlugin()
-    course_copy: CourseCopyPlugin = _LazyPlugin()
-    misc: MiscPlugin = _LazyPlugin()
+    grouping: GroupingPlugin = _LazyPlugin()  # type: ignore
```
Comment (marcospri) [inline lms/product/plugin/plugin.py:34]: This code it's a bit magical. Not worth trying to fix this.

File: lms/services/canvas_studio.py [hunk 3]
```diff
@@ -219,7 +219,7 @@ def list_collection(self, collection_id: str) -> list[File]:
 
         return files
 
-    def _api_request(self, path: str, schema_cls: RequestsResponseSchema) -> dict:
+    def _api_request(self, path: str, schema_cls: Type[RequestsResponseSchema]) -> dict:
```
Comment (marcospri) [inline lms/services/canvas_studio.py:222]: Object type vs passing a class.

File: lms/services/http.py [hunk 4]
```diff
@@ -7,7 +7,7 @@ class HTTPService:
     """Send HTTP requests with `requests` and receive the responses."""
 
     # This is here mostly to let auto-spec know about it in the tests
-    session: Session = None
+    session: Session
```
Comment (marcospri) [inline lms/services/http.py:10]: No need for the useless default.

File: lms/services/upsert.py [hunk 5]
```diff
@@ -45,12 +45,12 @@ def bulk_upsert(
 
     index_elements_columns = [getattr(model_class, c) for c in index_elements]
 
-    stmt = insert(model_class).values(values)
-    stmt = stmt.on_conflict_do_update(
+    base = insert(model_class).values(values)
```
Comment (marcospri) [inline lms/services/upsert.py:48]: These two had different types, use different names.

File: lms/services/vitalsource/service.py [hunk 6]
```diff
@@ -92,9 +92,9 @@ def get_document_url(
         url = urlparse(url)
         params = parse_qs(url.query)
         if end_page:
-            params["end_page"] = end_page
+            params["end_page"] = end_page  # type: ignore
```
Comment (marcospri) [inline lms/services/vitalsource/service.py:95]: parse_qs seems to return `dict[str, list[str]]`

File: lms/validation/_base.py [hunk 7]
```diff
@@ -12,11 +12,11 @@
 class PlainSchema(marshmallow.Schema):
     """Base class for all schemas."""
 
-    many = None
+    many: bool = False
     """
     Whether or not this schema validates collections of objects by default.
 
-    If this is ``None`` then marshmallow's default behavior will be used -- the
+    If this is ``False`` then marshmallow's default behavior will be used -- the
```
Comment (marcospri) [inline lms/validation/_base.py:19]: https://github.com/marshmallow-code/marshmallow/blob/dev/src/marshmallow/schema.py#L369

File: pyproject.toml [hunk 8]
```diff
@@ -147,7 +147,6 @@ python_version = 3.11
 
 disable_error_code = [
     "arg-type",
-    "assignment",
```
Comment (marcospri) [inline pyproject.toml:150]: The plan is to remove codes from here one by one.

File: lms/services/http.py [hunk 9]
```diff
@@ -7,7 +7,7 @@ class HTTPService:
     """Send HTTP requests with `requests` and receive the responses."""
 
     # This is here mostly to let auto-spec know about it in the tests
-    session: Session = None
+    session: Session = None  # typing: ignore
```
Comment (marcospri) [inline lms/services/http.py:10]: Removing the default makes the test fail, which is the reason to have this in the first place.

File: lms/product/blackboard/product.py [hunk 10]
```diff
@@ -3,14 +3,14 @@
...
 from lms.product.blackboard._plugin.misc import BlackboardMiscPlugin
-from lms.product.product import PluginConfig, Product, Routes
+from lms.product.product import Family, PluginConfig, Product, Routes
 
 
 @dataclass
 class Blackboard(Product):
     """A product for Blackboard specific settings and tweaks."""
 
-    family: Product.Family = Product.Family.BLACKBOARD
```
Comment (marcospri) [inline lms/product/blackboard/product.py:13]: Access these more directly

File: lms/product/plugin/plugin.py (hunk 2 above)
Comment (acelaya) [reply lms/product/plugin/plugin.py:34]: Should the comment be `# type:ignore`?

```suggestion
    grouping: GroupingPlugin = _LazyPlugin()  # type:ignore
```

File: lms/services/http.py (hunk 9 above)
Comment (robertknight) [reply lms/services/http.py:10]: I had a look through the Python docs for `create_autospec`, which suggests some alternative ways of handling this. I include the diff below to illustrate, but `# typing: ignore` is simpler.

```diff
diff --git a/lms/services/http.py b/lms/services/http.py
index 8d0e73fca..8c282b3c5 100644
--- a/lms/services/http.py
+++ b/lms/services/http.py
@@ -6,8 +6,7 @@ from lms.services.exceptions import ExternalRequestError
 class HTTPService:
     """Send HTTP requests with `requests` and receive the responses."""
 
-    # This is here mostly to let auto-spec know about it in the tests
-    session: Session = None  # typing: ignore
+    session: Session
     """The underlying requests Session."""
 
     def __init__(self):
@@ -15,7 +14,6 @@ class HTTPService:
         # requests and urllib3 connection pooling is used (which means that
         # underlying TCP connections are re-used when making multiple requests
         # to the same host, e.g. pagination).
-
         # See https://docs.python-requests.org/en/latest/user/advanced/#session-objects
         self.session = Session()
 
diff --git a/tests/unit/lms/services/jstor/service_test.py b/tests/unit/lms/services/jstor/service_test.py
index d5152d0c3..ea4d86ad1 100644
--- a/tests/unit/lms/services/jstor/service_test.py
+++ b/tests/unit/lms/services/jstor/service_test.py
@@ -5,7 +5,7 @@ from unittest.mock import sentinel
 import pytest
 
 from lms.services import ExternalRequestError
-from lms.services.jstor.service import ArticleNotFound, JSTORService
+from lms.services.jstor.service import ArticleNotFound, HTTPService, JSTORService
 from tests import factories
 
 API_URL = "http://jstore_api.example.com"
@@ -200,7 +200,12 @@ class TestJSTORService:
 
     @pytest.fixture(autouse=True)
     def HTTPService(self, patch):
-        return patch("lms.services.jstor.service.HTTPService")
+        class HTTPServiceForTest(HTTPService):
+            session = None
+
+        return patch(
+            "lms.services.jstor.service.HTTPService", autospec=HTTPServiceForTest
+        )
 
     @pytest.fixture(autouse=True)
     def JWTService(self, patch):
diff --git a/tests/unit/lms/services/vitalsource/_client_test.py b/tests/unit/lms/services/vitalsource/_client_test.py
index a0c43921e..9da609b36 100644
--- a/tests/unit/lms/services/vitalsource/_client_test.py
+++ b/tests/unit/lms/services/vitalsource/_client_test.py
@@ -7,6 +7,7 @@ from requests import Request
 from lms.services.exceptions import ExternalRequestError
 from lms.services.vitalsource._client import (
     BookNotFound,
+    HTTPService,
     VitalSourceClient,
     VitalSourceConfigurationError,
     _VSUserAuth,
@@ -253,9 +254,13 @@ class TestVitalSourceClient:
 
     @pytest.fixture(autouse=True)
     def http_service(self, patch):
-        HTTPService = patch("lms.services.vitalsource._client.HTTPService")
+        class HTTPServiceForTest(HTTPService):
+            session = None
 
-        return HTTPService.return_value
+        cls = patch(
+            "lms.services.vitalsource._client.HTTPService", autospec=HTTPServiceForTest
+        )
+        return cls.return_value
 
     @pytest.fixture
     def _VSUserAuth(self, patch):
diff --git a/tests/unit/services.py b/tests/unit/services.py
index 6e71bc381..a43513842 100644
--- a/tests/unit/services.py
+++ b/tests/unit/services.py
@@ -1,6 +1,7 @@
 from unittest import mock
 
 import pytest
+from requests import Session
 
 from lms.product.plugin.course_copy import (
     CourseCopyFilesHelper,
@@ -234,9 +235,11 @@ def h_api(mock_service):
 
 @pytest.fixture
 def http_service(mock_service):
-    http_service = mock_service(HTTPService, service_name="http")
-    http_service.request.return_value = factories.requests.Response()
+    class HTTPServiceForTest(HTTPService):
+        session = mock.create_autospec(Session, spec_set=True, instance=True)
 
+    http_service = mock_service(HTTPServiceForTest, service_name="http")
+    http_service.request.return_value = factories.requests.Response()
     return http_service
```

**Review by robertknight — APPROVED** [review]
From the changes here the one thing I think we might want to look into a better pattern for is avoiding adding extra `| None`s for dataclass fields which are initialized in `__post_init__` since that will lead to unnecessary checks in code that handles constructed dataclass instances.

File: lms/services/http.py (hunk 9 above)
Comment (acelaya) [reply lms/services/http.py:10]: Should we then define the type as `Session | None`?

**Review by acelaya — APPROVED** [review]

File: lms/product/plugin/plugin.py (hunk 2 above)
Comment (marcospri) [reply lms/product/plugin/plugin.py:34]: Went for the type casting :+1:

File: lms/config.py [hunk 11]
```diff
@@ -39,8 +39,8 @@ class _Setting:
     """The properties of a setting and how to read it."""
 
     name: str
-    read_from: str = None
-    value_mapper: Callable = None
+    read_from: str | None = None
```
[the remark this answers is not in the document]
Comment (marcospri) [reply lms/config.py:42]: I'll keep looking for a good solution in a different PR.

The feature we are after is for dataclass fields that can be used during __init__ but post__init provides a default if that's missing.

**Pull request merged**